prql-compiler/src/span.rs

```diff
@@ -0,0 +1,128 @@
+
+    fn add(self, rhs: usize) -> Span {
+        Span {
+            start: self.start + rhs,
+            end: self.end + rhs,
+            source_id: self.source_id,
+        }
+    }
+}
+
+impl Debug for Span {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{}-{}", self.source_id, self.start, self.end)
+    }
+}
+
+impl Serialize for Span {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let str = format!("{self:?}");
+        serializer.serialize_str(&str)
+    }
+}
+
+impl<'de> Deserialize<'de> for Span {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct SpanVisitor {}
+
+        impl<'de> Visitor<'de> for SpanVisitor {
+            type Value = Span;
+
+            fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                write!(f, "A span string of form `file_id:x-y`")
+            }
+
+            fn visit_str<E>(self, v: &str) -> std::result::Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                use serde::de;
+
+                if let Some((file_id, char_span)) = v.split_once(':') {
+                    let file_id = file_id
+                        .parse::<u16>()
+                        .map_err(|e| de::Error::custom(e.to_string()))?;
+
+                    if let Some((start, end)) = char_span.split_once('-') {
+                        let start = start
+                            .parse::<usize>()
+                            .map_err(|e| de::Error::custom(e.to_string()))?;
+                        let end = end
+                            .parse::<usize>()
+                            .map_err(|e| de::Error::custom(e.to_string()))?;
+
+                        return Ok(Span {
+                            start,
+                            end,
+                            source_id: file_id,
+                        });
+                    }
+                }
+
+                Err(de::Error::custom("malformed span"))
+            }
+
+            fn visit_string<E>(self, v: String) -> std::result::Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                self.visit_str(&v)
+            }
+        }
+
+        deserializer.deserialize_string(SpanVisitor {})
+    }
+}
+
+impl chumsky::Span for Span {
+    type Context = u16;
+
+    type Offset = usize;
+
+    fn new(context: Self::Context, range: std::ops::Range<Self::Offset>) -> Self {
+        Span {
+            start: range.start,
+            end: range.end,
+            source_id: context,
+        }
+    }
+
+    fn context(&self) -> Self::Context {
+        self.source_id
+    }
+
+    fn start(&self) -> Self::Offset {
+        self.start
+    }
+
+    fn end(&self) -> Self::Offset {
+        self.end
+    }
+}
```

Comment:
Let's keep this in the parser module, as it is very parser specific. If we didn't want to be implementing traits so far from the type declaration, we could create a newtype for the parser span.

Reply:
I have now actually moved the `span` module back into the root of the crate where I had it originally and just kept the `chumsky::Span` impl in the parser module as you suggested. The reasoning being that in the followup PR I'll make the parser generic over the used Span type so it wouldn't make sense to have the concrete type definition residing in the parser module.

Reply:
Nevermind ... I looked into it and making the parser generic over `chumsky::Span` is actually quite cumbersome since a bunch of code in the parser currently does some arithmetic on the `usize`s. So in the followup PR I'll just restrict the associated type of the trait to the `Span` from the parser module.

Reply:
> make the parser generic over the used Span

Wait, why? Can we not make an assumption that we are parsing text files, where span is a range of chars?

> So in the followup PR I'll just restrict the associated type of the trait to the Span from the parser module.

What do you mean here?

Reply:
It's not really important. It just would be nice if you could use e.g. `()` as the span type if you don't care about spans at all but yeah that's more of a micro optimization ... not important.

>What do you mean here?

If you look at the parser files in #2994 now you'll see a bunch of `<T: Extension<Span = Span>>` for every method, which says if you want to use the parser the Span has to be the span from the parser crate. But yeah I think that's actually fine ... the important improvement is that the Span stays with the parser and is no longer moved to the new prql_ast crate as I did initially.

Reply:
Ok, yes, `<T: Extension<Span = Span>>` just means that there is a dependency from our parser to chumsky. Which is completely ok.

Reply:
I mean it of course depends on chumsky since it's implemented with chumsky ... a chumsky parser could in theory be implemented to work with an arbitrary span type that implements `chumsk::Span` ... `<T: Extension<Span = Span>>` constraints the span type supported by our parser to our own span type ... but yeah that constraint already exists anyway.

Reply:
Wait, so why is span now under `parser` module? That should be under AST, as the AST uses the span?

Reply:
The prql_ast crate introduced by the followup PR is generic over the span. We'll want to split off the `parser` module from `prql-compiler` in the future at which point `prql-compiler` will depend on `prql-parser` but the `Span` type will be part of `prql-parser` since you cannot use the parser without it.

Reply:
Let's keep the span under root module or ast module for now, so this can be merged.

Reply:
I can do that but I am curious on why you prefer that. I think it's clear that it has the tightest dependency with the parser and extracting the parser into its own crate will require the Span to be moved along.

Reply:
I see the span as a part of AST, and the parser is just something that produces the span. This is not the only interpretation, but to me it makes the most sense.

Reply:
When we extract the parser it would depend on ast crate the same way as it depends on ast crate for other AST nodes.

Reply:
Ok yeah that makes sense.